dali/operators/reader/video_reader_decoder_gpu_op.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "dali/operators/reader/video_reader_decoder_gpu_op.h"
+
+#include <string>
+#include <vector>
+
+namespace dali {
+
+VideoReaderDecoderGpu::VideoReaderDecoderGpu(const OpSpec &spec)
+    : DataReader<GPUBackend, VideoSampleGpu>(spec),
+      has_labels_(spec.HasArgument("labels")) {
+      loader_ = InitLoader<VideoLoaderDecoderGpu>(spec);
+}
+
+void VideoReaderDecoderGpu::PrepareOutput(TensorList<GPUBackend> &video_output) {
+}
+
+void VideoReaderDecoderGpu::RunImpl(DeviceWorkspace &ws) {
+  auto &video_output = ws.Output<GPUBackend>(0);
+  auto &current_batch = prefetched_batch_queue_[curr_batch_consumer_];
+  int batch_size = current_batch.size();
+
+  TensorListShape<4> output_shape;
+  output_shape.resize(batch_size);
+
+  for (int sample_id = 0; sample_id < batch_size; ++sample_id) {
+    auto &sample = current_batch[sample_id];
+    output_shape.set_tensor_shape(sample_id, sample->data_.shape());
+  }
+
+  video_output.Resize(output_shape, current_batch[0]->data_.type());
```

Comment:
I moved decoding the the RunImpl. This means that we do not have the shape of the output ready during `SetupImpl`.
We could pass the size from the `FramesDecoder`, through `VideoSampleDesc` but I don't want to do this as this will limit us in the future, when we optimize index building and might not know size without decoding.
When this is more or less feature complete, we can revisit this refactoring, ok?